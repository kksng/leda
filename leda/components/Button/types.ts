import React from 'react';
import { COMPONENTS_NAMESPACES } from '../../constants';
import { PartialGlobalDefaultTheme } from '../../utils/useTheme';
import { Form, FormsObject, ValidationButtonProps } from '../Validation/types';

export interface SubmitEvent extends React.MouseEvent<HTMLButtonElement> {
  form?: FormsObject,
  forms?: Form[],
}

export type ValidationFailEvent = React.MouseEvent<HTMLButtonElement> & { invalidForms: Form[] };

export interface ButtonProps extends React.HTMLAttributes<HTMLButtonElement>, ValidationButtonProps {
  /** Название формы */
  form?: string | string[],
  /** Отключение кнопки */
  isDisabled?: boolean,
  /** Показывать лоадер */
  isLoading?: boolean,
  /** Обработчик клика */
  onClick?: (event: SubmitEvent) => void,
  /** Применяется к кнопке. Обработчик, который срабатывает, если есть невалидные поля,
   * если срабатывает onValidationFail, то onClick на кнопке не сработает */
  onValidationFail?: (ev: ValidationFailEvent) => void,
  /** Реф */
  ref?: React.Ref<ButtonRefCurrent>,
  /** Задержкаа перед скроллм к невалидным полям */
  scrollDelay?: number,
  /** Дополнительный сдвиг в px при скролинге к кнопке */
  scrollOffset?: number,
  /** Если передано false, компонент не будет рендериться */
  shouldRender?: boolean,
  /** Прокрутка к невалидным полям при нажатии на кнопку */
  shouldScrollToInvalidFields?: boolean,
  /** Тема компонета */
  theme?: PartialGlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.button],
  /** Тип кнопки */
  type?: React.ButtonHTMLAttributes<{}>['type'],
  /** Классы переданные через _ */
  [x: string]: unknown,
}

export interface ButtonRefCurrent {
  wrapper: HTMLButtonElement | null,
}
