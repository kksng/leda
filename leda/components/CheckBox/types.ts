import * as React from 'react';
import { CustomRender } from '../../commonTypes';
import { PartialGlobalDefaultTheme } from '../../utils/useTheme';
import { COMPONENTS_NAMESPACES } from '../../constants';

export interface ChangeEvent extends React.ChangeEvent<HTMLInputElement> {
  component: {
    value: boolean,
    name?: string,
  },
}

export interface CheckBoxProps {
  /** Всё, что обёрнуто в L.Checkbox, попадает в label */
  children?: React.ReactNode,
  /** Значение по-умолчанию, если не передано - false */
  defaultValue?: boolean,
  /** Id для чекбокса */
  id?: string,
  /** Кастомизация инпута, непосредственно инпут - невидим, но данный рендер позволяет добавить атрибуты в тег <input> */
  inputRender?: CustomRender<CheckBoxProps, {}, React.InputHTMLAttributes<HTMLInputElement>>,
  /** Отключение чекбокса */
  isDisabled?: boolean,
  
  /** Флаг частичного выбора */
  isSemi?: boolean,

  /** Кастомизация элемента label */
  labelRender?: CustomRender<CheckBoxProps, {}, React.LabelHTMLAttributes<any>>,
  /** Имя компонента */
  name?: string,
  /** Обработчик изменения состояния элементов */
  onChange?: (event: ChangeEvent) => void,
  /** Реф */
  ref?: React.Ref<CheckBoxRefCurrent>,
  /** Тема компонента */
  theme?: PartialGlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.checkBox],
  /** Значение чекбокса */
  value?: boolean,
  /** Кастомизация враппера */
  wrapperRender?: CustomRender<CheckBoxProps, {}, Partial<CheckBoxProps>>,
  /** Классы переданные через _ */
  [x: string]: unknown,
}

export interface CheckBoxRefCurrent {
  wrapper: HTMLElement | null,
  input: HTMLInputElement | null,
  label: HTMLLabelElement | null,
}
