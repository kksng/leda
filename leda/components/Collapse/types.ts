import * as React from 'react';
import { CustomRender, CustomEventHandler } from '../../commonTypes';
import { COMPONENTS_NAMESPACES } from '../../constants';
import { GlobalDefaultTheme, PartialGlobalDefaultTheme } from '../../utils/useTheme';

export interface SelectEvent {
  component: {
    value: string | string[] | null,
  },
}

export interface BodyClickCustomEvent {
  component: {
    value: string, // panelKey
  },
}


export interface CollapseProps {
  /** Признак активной панели. Сюда передается значение panelKey активной панели. Когда панель(панели) закрыта, значение равно null */
  activePanelKey?: string | string[] | null,
  /** Дочерние элементы */
  children: React.ReactNode,
  /** Признак возможности открытия нескольких панелей для неконтролируемого режима */
  isAccordion?: boolean,
  /** Обработчик клика на панель */
  onSelect?: CustomEventHandler<SelectEvent>,
  /** Тема */
  theme?: PartialGlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.collapse],
}

export interface HeadingProps {
  /** Дочерние элементы */
  children: React.ReactNode,
  /** Иконка возле заголовка */
  iconRender?: CustomRender<HeadingProps, { }, IconProps>,
  /** Обработчик клика */
  onClick?: React.MouseEventHandler<HTMLDivElement>,
  /** Реф */
  ref?: React.Ref<HeadingRefCurrent>,
  /** Компонент-обертка в который будет помещен Collapse.Heading. Передавать в виде <Wrapper props />. По умолчанию - <Div /> */
  wrapperRender?: CustomRender<HeadingProps, { }, HeadingWrapperProps>,
  /** Классы переданные через _ */
  [x: string]: unknown,
}

export interface BodyProps {
  /** Дочерние элементы */
  children?: React.ReactNode,
  /** Состояние загрузки данных */
  isLoading?: boolean,
  /** Функция обратного вызова на закрытие коллапса */
  onClose?: CustomEventHandler<BodyClickCustomEvent>,
  /** Функция обратного вызова на закрытие коллапса по клику */
  onCloseByClick?: CustomEventHandler<BodyClickCustomEvent>,
  /** Функция обратного вызова на открытие коллапса */
  onOpen?: CustomEventHandler<BodyClickCustomEvent>,
  /** Реф */
  ref?: React.Ref<BodyRefCurrent>,
  /** Описание открытия/закрытия коллапса. Описывается как CSS Transition. См. https://developer.mozilla.org/ru/docs/Web/CSS/transition. По умолчанию height 250ms cubic-bezier(.4, 0, .2, 1) */
  transition?: string,
  /** Компонент-обертка в который будет помещен Collapse.Body. Передавать в виде <Wrapper props />. По умолчанию - <Div /> */
  wrapperRender?: CustomRender<BodyProps, { }, BodyWrapperProps>,
  /** Классы переданные через _ */
  [x: string]: unknown,
}

export interface PanelProps {
  /** Дочерние элементы */
  children: React.ReactNode,
  /** Отключение панели */
  isDisabled?: boolean,
  /** Имя панели */
  name?: string,
  /** Идентификатор события панели, данное значение передается в activeKey компонента Collapse */
  panelKey: string,
  /** Компонент-обертка для панели, например Div. Передавать в виде <Wrapper props /> */
  wrapperRender?: CustomRender<PanelProps, { isClicked: boolean }, PanelWrapperProps>,
  /** Классы переданные через _ */
  [x: string]: unknown,
}

export interface CollapseContextType {
  activePanelKey: string | string[] | null,
  onSelect: SelectHandler,
  theme: GlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.collapse],
}

export interface CollapsePanelContextType extends CollapseContextType {
  panelKey: string,
  isClicked: boolean,
  isDisabled?: boolean,
  name?: string,
  isExpanded: boolean,
  onHeadingClick: CollapsePanelHandlers['onHeadingClick'],
  onBodyRest: CollapsePanelHandlers['onBodyRest'],
}

export interface HeadingWrapperProps {
  children?: React.ReactNode,
  className?: string,
  onClick: React.MouseEventHandler,
}

export interface IconProps {
  className?: string,
  isExpanded: boolean,
}

export interface BodyWrapperProps {
  className?: string,
  children?: React.ReactNode,
}

export interface PanelWrapperProps {
  children?: React.ReactNode,
  className?: string,
}

export type KeyState = string | string[] | null;

export interface SelectHandler {
  (ev: BodyClickCustomEvent): void,
}

export type SetClicked = React.Dispatch<React.SetStateAction<boolean>>;

export interface CollapsePanelHandlers {
  onHeadingClick: () => void,
  onBodyRest: () => void,
}

export type CollapseComponent = React.FC<CollapseProps> & { Panel: React.FC<PanelProps>, Body: React.FC<BodyProps>, Heading: React.FC<HeadingProps> };

export interface BodyRefCurrent {
  wrapper: HTMLElement | null,
}

export interface HeadingRefCurrent {
  wrapper: HTMLElement | null,
}
