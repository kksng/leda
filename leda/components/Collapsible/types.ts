import * as React from 'react';

export interface TransitionOptions {
  /** Длительность анимации коллапса. Число в миллисекундах. Строки с единицами измерения */
  duration: number | string,
  /** Функция анимации коллапса. См. https://developer.mozilla.org/en-US/docs/Web/CSS/transition-timing-function */
  animation: string,
  /** Длительность задержки перед началом анимации. Число в миллисекундах. Строки с единицами измерения */
  delay?: number | string,
}

export interface CollapsibleProps extends React.HTMLAttributes<HTMLDivElement> {
  /** Состояние элемента открыт/закрыт */
  isOpen: boolean,
  /** Функция обратного вызова при окончании закрытия коллапса */
  onClose?: () => void,
  /** Функция обратного вызова при окончании открытия коллапса */
  onOpen?: () => void,
  /** Функция обратного вызова при окончании анимации коллапса */
  onToggle?: () => void,
  /** Реф */
  ref?: React.Ref<CollapsibleRefCurrent>,
  /** Описание открытия/закрытия коллапса. Описывается как CSS Transition. См. https://developer.mozilla.org/en-US/docs/Web/CSS/transition. По умолчанию height 250ms cubic-bezier(.4, 0, .2, 1) */
  transition?: string | TransitionOptions,
}

export interface CollapsibleRefCurrent {
  wrapper: HTMLDivElement | null,
}
