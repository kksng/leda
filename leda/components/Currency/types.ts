import * as React from 'react';
import { CustomRender } from '../../commonTypes';

export interface CurrencyProps {
  /** значение, как потомок */
  children?: number | string,
  /** Валюта в формате ISO 4217 - EUR, USD, RUB etc, по-умолчанию RUB */
  currencyCode?: string,
  /** Кастомный рендер знака валюты (для добавления единиц измерения, например) */
  currencySymbolRender?: CustomRender<CurrencyProps, {}, CurrencySymbolProps>,
  /** что отображать, если нет значения */
  placeholder?: string,
  /** кол-во знаков после запятой */
  precision?: number,
  /** Реф */
  ref?: React.Ref<CurrencyRefCurrent>,
  /** Убирает дробную часть для целых чисел, игнорируя precision */
  shouldTrimFraction?: boolean,
  /** значение поля */
  value?: number | string,
  /** Кастомизация враппера (по-умолчанию span) */
  wrapperRender?: CustomRender<CurrencyProps, {}, WrapperProps>,
  /** Классы переданные через _ */
  [x: string]: unknown,
}

export interface CurrencySymbolProps {
  style?: Partial<CSSStyleDeclaration>,
}

export interface WrapperProps {
  ref?: React.Ref<CurrencyRefCurrent>,
}

export interface CurrencyRefCurrent {
  wrapper: HTMLElement | null,
}
