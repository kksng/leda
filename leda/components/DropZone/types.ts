import * as React from 'react';
import { CustomRender, CustomEventHandler } from '../../commonTypes';
import { PartialGlobalDefaultTheme } from '../../utils/useTheme';
import { COMPONENTS_NAMESPACES } from '../../constants';
import { globalDefaultTheme } from '../LedaProvider';
import { ValidationProps } from '../Validation/types';

export { FileErrorCodes } from '../../constants';

export interface ExternalFile {
  /** Имя файла. Необходимо для отображения в списке и удаления */
  name: string,
  /** Ссылка на скачивание файла. При наличии, файл будет отображен в списке, как скачиваемый */
  link?: string,
  /** Синоним имени файла. Необходимо для отображения в списке и удаления */
  path?: string,
  /** Тип файла */
  type?: string,
}

export interface DropZoneFileType extends File {
  path?: string,
  preview?: string,
  lastModified: number,
  errorCode?: number,
}

export type FileType = ExternalFile | DropZoneFileType;

export interface ChangeEvent {
  component: {
    dropped: {
      acceptedFiles: FileType[],
      rejectedFiles: FileType[],
    },
    value: {
      acceptedFiles: FileType[],
      rejectedFiles: FileType[],
    },
    removedFile?: FileType,
  },
}

export interface DropZoneProps extends ValidationProps {
  /** Отображение добавленных файлов */
  acceptedFilesRender?: CustomRender<DropZoneProps, DropZoneState, AcceptedFilesProps>,
  /** Разрешенные типы файлов, см. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#Attributes. Передача нескольких типов файлов происходит через запятую (.png, image/jpeg). allowedFiles и forbiddenFiles вместе не могут находиться. */
  allowedFiles?: string,
  /** Классы, применяемые к компоненту */
  className?: string,
  /** DOM-нода в которой будет отрендерен список файлов */
  dropZoneFilesNode?: Element,
  /** Список загруженных файлов */
  value?: {
    acceptedFiles: ExternalFile[],
    rejectedFiles: ExternalFile[],
  },
  /** Запрещенные типы файлов. см. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#Attributes. Передача нескольких типов файлов происходит через запятую (.png, image/jpeg). allowedFiles и forbiddenFiles вместе не могут находиться. */
  forbiddenFiles?: string,
  /** Кастомизация описания компонента */
  infoRender?: CustomRender<DropZoneProps, DropZoneState, InfoProps>,
  /** Признак отключения дропзоны */
  isDisabled?: boolean,
  /* Максимальная длина имени файла, по-умолчанию 255 символов */
  maxFileNameLength?: number,
  /** Максимальное количество файлов */
  maxFilesNumber?: number,
  /** Максимальный размер файла, в байтах */
  maxFileSize?: number,
  /** Минимальный размер файла, в байтах */
  minFileSize?: number,
  /** Функция обратного вызова для метода onChange */
  onChange?: (event: ChangeEvent) => void,
  /** Отображение отклоненных файлов */
  rejectedFilesRender?: CustomRender<DropZoneProps, DropZoneState, RejectedFilesProps>,
  ref?: React.Ref<DropZoneRefCurrent>,
  /* Тема для компонента */
  theme?: PartialGlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.dropZone],
  /** Текст кнопки загрузки файла, может принимать JSX */
  uploadButtonRender?: CustomRender<DropZoneProps, DropZoneState, UploadButtonProps>,
  /** Кастомизация враппера */
  wrapperRender?: CustomRender<DropZoneProps, DropZoneState, WrapperProps>,
  /** Классы переданные через _ */
  [x: string]: unknown,
}

export interface DropZoneState {
  /* список принятых файлов */
  acceptedFiles: DropZoneFileType[],
  /* список файлов, не прошедших валидацию по типу или размеру */
  rejectedFiles: DropZoneFileType[],
}

export interface AcceptedFilesProps {
  children?: React.ReactNode,
  dropZoneFilesNode?: Element,
  onChange?: ChangeEventHandler,
}

export interface RejectedFilesProps {
  children?: React.ReactNode,
  className?: string,
}

export interface UploadButtonProps {
  children?: React.ReactNode,
  className?: string,
  onClick?: CustomEventHandler<React.MouseEvent>,
  [x: string]: unknown,
}

export interface InfoProps {
  children?: React.ReactNode,
  className?: string,
}

export interface WrapperProps {
  children?: React.ReactNode,
  className?: string,
  ref: React.Ref<DropZoneRefCurrent>,
}

export interface CustomElements {
  AcceptedFiles: React.FC<AcceptedFilesProps>,
  RejectedFiles: React.FC<RejectedFilesProps>,
  Info: React.FC<InfoProps>,
  UploadButton: React.FC<UploadButtonProps>,
  Wrapper: React.FC<WrapperProps>,
}

export interface DropZoneFilesProps {
  /** Массив файлов, который будет отображен в виде списка */
  files: ExternalFile[],
  /** Обработчик удаления файла */
  onChange: (
    acceptedFiles: ExternalFile[] | DropZoneFileType[],
    rejectedFiles: ExternalFile[] | DropZoneFileType[],
    ev: React.DragEvent<HTMLDivElement> | React.MouseEvent<HTMLAnchorElement>,
    removedFile: ExternalFile | DropZoneFileType
  ) => void,
  /** Флаг обозначающий нужно ли отображать файлы */
  shouldRender: boolean,
  /** Принятые и отклоненные файлы */
  value: DropZoneState | NonNullable<DropZoneProps['value']>,
  /** Тема компонента */
  theme: typeof globalDefaultTheme[typeof COMPONENTS_NAMESPACES.dropZone],
}

export interface DropZoneError {
  message: string,
  errorCode: number,
}

export interface RejectedFilesListProps {
  theme: typeof globalDefaultTheme[typeof COMPONENTS_NAMESPACES.dropZone],
  value: DropZoneState | NonNullable<DropZoneProps['value']>,
  maxFilesNumber?: number,
}

export interface ChangeEventHandler {
  (
    accepted: FileType[],
    rejected: FileType[],
    ev?: React.DragEvent<HTMLElement> | React.MouseEvent<HTMLElement>,
    removedFile?: FileType
  ): void,
}

export interface DropZoneRefCurrent {
  wrapper: HTMLElement | null,
  input: HTMLInputElement | null,
}
