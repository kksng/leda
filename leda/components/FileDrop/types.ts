import * as React from 'react';
import { CustomRender, CustomEventHandler } from '../../commonTypes';
import { GlobalDefaultTheme, PartialGlobalDefaultTheme } from '../../utils/useTheme';
import { COMPONENTS_NAMESPACES } from '../../constants';
import { ValidationProps } from '../Validation/types';

export interface FileType extends File {
  /** Код ошибки, подробнее можно посмотреть в leda/constants.ts */
  errorCode?: number,
  /** Сообщение об ошибке */
  errorMessage?: string,
  /** Дата последнего изменения */
  lastModified: number,
  /** Ссылка на скачивание файла. При наличии, файл будет отображен в списке, как скачиваемый */
  link?: string,
  /** Имя файла. Необходимо для отображения в списке и удаления */
  name: string,
  /** Синоним имени файла. Необходимо для отображения в списке и удаления */
  path?: string,
  /** Предварительный просмотр */
  preview?: string,
  /** Тип файла */
  type: string,
}

export interface ChangeEvent {
  component: {
    value: FileType | null,
  },
}

export interface LoadingData {
  error?: Error | null,
  loaded: number,
  total: number,
}

export interface FileDropProps extends ValidationProps {
  /** Разрешенные типы файлов, см. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#Attributes. Передача нескольких типов файлов происходит через запятую (.png, image/jpeg). allowedFiles и forbiddenFiles вместе не могут находиться. */
  allowedFiles?: string,
  /** Классы, применяемые к компоненту */
  className?: string,
  /** Список загруженных файлов */
  value: FileType | null,
  /** Запрещенные типы файлов. см. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#Attributes. Передача нескольких типов файлов происходит через запятую (.png, image/jpeg). allowedFiles и forbiddenFiles вместе не могут находиться. */
  forbiddenFiles?: string,
  /** Кастомизация описания компонента */
  infoRender?: CustomRender<FileDropProps, {}, InfoProps>,
  /** Признак отключения дропзоны */
  isDisabled?: boolean,
  /** Данные о загрузке, используется для отображения прогресса, чтобы получить эти данные используйте XMLHTTPRequest */
  loadingData?: LoadingData | null,
  /* Максимальная длина имени файла, по-умолчанию 255 символов */
  maxFileNameLength?: number,
  /** Максимальный размер файла, в байтах */
  maxFileSize?: number,
  /** Минимальный размер файла, в байтах */
  minFileSize?: number,
  /** Функция обратного вызова для метода onChange */
  onChange: (event: ChangeEvent) => void,
  /** Функция-обработчик onClick */
  onClick?: (event: React.MouseEvent) => void,
  /** Реф */
  ref?: React.Ref<FileDropRefCurrent>,
  /** Тема для компонента */
  theme?: PartialGlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.fileDrop],
  /** Текст кнопки загрузки файла, может принимать JSX */
  uploadButtonRender?: CustomRender<FileDropProps, {}, UploadButtonProps>,
  /** Кастомизация враппера */
  wrapperRender?: CustomRender<FileDropProps, {}, WrapperProps>,
  /** Классы переданные через _ */
  [x: string]: unknown,
}

export interface UploadButtonProps {
  children?: React.ReactNode,
  className?: string,
  onClick?: CustomEventHandler<React.MouseEvent>,
  [x: string]: unknown,
}

export interface InfoProps {
  children?: React.ReactNode,
  className?: string,
}

export interface WrapperProps {
  children?: React.ReactNode,
  className?: string,
  ref: React.Ref<FileDropRefCurrent>,
}

export interface CustomElements {
  Info: React.FC<InfoProps>,
  UploadButton: React.FC<UploadButtonProps>,
  Wrapper: React.FC<WrapperProps>,
}

export interface FileDropError {
  message: string,
  errorCode: number,
}

export interface ChangeEventHandler {
  (
    accepted: FileType[],
    rejected: FileType[],
    ev?: React.DragEvent<HTMLDivElement> | React.MouseEvent<HTMLAnchorElement>,
    removedFile?: FileType
  ): void,
}

export interface FileDropRefCurrent {
  wrapper: HTMLElement | null,
  input: HTMLInputElement | null,
}

export interface ProgressLoaderProps {
  loadingData?: LoadingData | null,
  isLoading: boolean,
  theme: GlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.fileDrop],
}

export interface SingleFileViewProps extends FileDropProps {
  theme: GlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.fileDrop],
  value: FileDropProps['value'],
  UploadButton: React.FC<UploadButtonProps>,
  Info: React.FC<InfoProps>,
  handleRetry: CustomEventHandler<React.MouseEvent<HTMLElement>>,
}
