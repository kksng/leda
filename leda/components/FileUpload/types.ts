import * as React from 'react';
import { RenderEvent } from '../../commonTypes';
import { ValidationProps } from '../Validation/types';
import { FileErrorCodes } from '../../constants';

export { FileErrorCodes } from '../../constants';

export interface RejectedFileType extends File {
  errorCode?: FileErrorCodes,
}

export interface FileLoadEvent {
  component: {
    value: {
      acceptedFiles: File[],
      rejectedFiles: RejectedFileType[],
    },
  },
}

export interface FileUploadError {
  /** Код ошибки, подробнее можно посмотреть в leda/constants.ts */
  errorCode: FileErrorCodes,
  /** Сообщение об ошибке */
  errorMessage: string,
}

export interface ChangeEvent {
  component: {
    error: FileUploadError | null,
    name?: string,
    value: File | null,
  },
}

export interface FileUploadProps extends ValidationProps {
  /** Разрешенные типы файлов, см. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#Attributes. Передача нескольких типов файлов происходит через запятую (.png, image/jpeg). allowedFiles и forbiddenFiles нельзя использовать вместе */
  allowedFiles?: string, // comma separated list of valid mime types: "image/*, application/pdf"
  /** Классы, применяемые к первому диву */
  className?: string,
  /** Описание компонента */
  infoRender?: (props: RenderEvent<FileUploadProps>) => React.ReactElement | React.FC,
  /** Состояние загрузки */
  isLoading?: boolean,
  /** Запрещенные типы файлов. см. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#Attributes. Передача нескольких типов файлов происходит через запятую (.png, image/jpeg). allowedFiles и forbiddenFiles нельзя использовать вместе */
  forbiddenFiles?: string,
  /** Максимальный размер файла в Мбайтах */
  maxFileSize?: number,
  /** Минимальный размер файла в байтах */
  minFileSize?: number,
  /** Обработчик клика */
  onClick?: (event: React.MouseEvent<HTMLDivElement>) => void,
  /** Функция обратного вызова. Получает в качстве аргументов принятые файлы и отклоненные файлы с кодом ошибки (1 - файл меньше минимального размера, 2 - больше максимального, 3 - не удовлетворяет типу, 0 - неизвестная ошибка) */
  onChange?: (event: ChangeEvent) => void,
  /** Функция обратного вызова. Получает в качстве аргументов принятые файлы и отклоненные файлы с кодом ошибки (1 - файл меньше минимального размера, 2 - больше максимального, 3 - не удовлетворяет типу, 0 - неизвестная ошибка) */
  onFileLoad?: (event: FileLoadEvent) => void,
  /** Реф */
  ref?: React.Ref<FileUploadRefCurrent>,
  /** Обертка компонента */
  wrapperRender?: (props: RenderEvent<FileUploadProps, FileUploadProps>) => React.ReactElement | React.FC,
  /** Классы объявленные с _ */
  [x: string]: unknown,
}

export interface WrapperProps {
  className?: string,
  ref: React.Ref<FileUploadRefCurrent>,
  onClick: (ev: React.MouseEvent<HTMLDivElement>) => void,
}

export interface InfoProps {
  children: React.ReactNode,
}

export interface CustomElements {
  Wrapper: React.FC<WrapperProps>,
  Info: React.FC<InfoProps>,
}

export interface FileUploadRefCurrent {
  wrapper: HTMLElement | null,
}
