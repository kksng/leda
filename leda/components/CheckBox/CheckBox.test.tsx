import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { CheckBox } from './index';
import { ChangeEvent } from './types';

describe('CheckBox SNAPSHOTS', () => {
  it('should render', () => {
    const { container } = render(<CheckBox id="test" />);

    expect(container.querySelector('input')).toBeDefined();

    expect(container.querySelector('label')).toBeDefined();

    expect(container).toMatchSnapshot();
  });

  it('should render controllable mode', () => {
    const eventMatcher = expect.objectContaining({
      component: expect.objectContaining({
        value: false,
        name: 'checker',
      }),
    });
    const Wrapper = () => {
      const [value, setValue] = React.useState(true);
      const handleChange = (ev: ChangeEvent) => {
        expect(ev).toMatchObject(eventMatcher);
        setValue(!value);
      };
      return (
        <CheckBox onChange={handleChange} id="test" name="checker" value={value} />
      );
    };
    const { container } = render(<Wrapper />);

    const input = document.querySelector('input');

    expect(input).toBeInTheDocument();

    expect(input?.checked).toBeTruthy();

    expect(container).toMatchSnapshot();

    const label = document.querySelector('label');

    expect(label).not.toEqual(null);

    userEvent.click(label as HTMLElement);

    expect(input?.checked).toBeFalsy();

    expect(container).toMatchSnapshot();
  });
});

describe('CheckBox HANDLERS', () => {
  it('should trigger onChange', () => {
    const onChange = jest.fn();
    const eventMatcher = expect.objectContaining({
      component: expect.objectContaining({
        value: true,
        name: 'chegevara',
      }),
    });
    const { container } = render(<CheckBox onChange={onChange} id="test" name="chegevara" />);

    const input = container.querySelector('input');

    expect(input).toBeInTheDocument();

    expect(input?.checked).toBeFalsy();

    userEvent.click(container.querySelector('label') as HTMLElement);

    expect(input?.checked).toBeTruthy();

    expect(onChange).toHaveBeenCalledTimes(1);

    expect(onChange).toHaveBeenCalledWith(eventMatcher);
  });

  describe('should trigger onClick', () => {
    it('should trigger onClick', () => {
      const onClick = jest.fn();

      const { container } = render(<CheckBox id="test" onClick={onClick} />);

      const input = container.querySelector('input');

      expect(input).toBeInTheDocument();

      expect(input?.checked).toBeFalsy();
      // кликнем по инпуту
      userEvent.click(container.querySelector('label') as HTMLElement);
      // проверим, что значение изменилось
      expect(input?.checked).toBeTruthy();

      // проверим, что был вызван обработик
      expect(onClick).toHaveBeenCalledTimes(1);
    });

    it('should not call onChange if checkbox is disabled', () => {
      const onChange = jest.fn();
      const { container } = render(<CheckBox isDisabled onChange={onChange} />);
      const input = container.querySelector('input');

      expect(input).toBeInTheDocument();

      expect(input?.disabled).toBeTruthy();
      // вызываем click
      userEvent.click(container.querySelector('label') as HTMLElement);
      // проверим, что значение в инпуте не изменилось
      expect(input?.checked).toBeFalsy();
      expect(onChange).not.toHaveBeenCalled();
    });
  });
});

describe('CheckBox ATTRIBUTES', () => {
  describe(', should add classname "semi" to wrapper', () => {
    it('should have className', () => {
      const { container } = render(<CheckBox _semi />);

      expect(container.querySelector('input')).toHaveClass('checkbox-input');

      expect(container.querySelector('label')).toHaveClass('checkbox-label');

      expect(container.querySelector('label.checkbox-label')).toHaveClass('semi');
    });

    it(', should convert props to classes', () => {
      const { container } = render(<CheckBox _active />);

      expect(container.querySelector('.checkbox-label.box')).toBeNull();

      expect(container.querySelector('.checkbox-label.active')).toBeInTheDocument();
    });

    it('classes passed through className should not override prop-classes', () => {
      const { container } = render(<CheckBox className="test" _active />);

      expect(container.querySelector('.checkbox-label.test')).toBeInTheDocument();

      expect(container.querySelector('.checkbox-label.active')).toBeInTheDocument();
    });
  });

  describe('children prop', () => {
    it('children should have correct className', () => {
      const { container } = render(<CheckBox><div className="lvl1"><span className="lvl2">TEXT</span></div></CheckBox>);

      expect(container.querySelector('div.lvl1')).toBeInTheDocument();

      expect(container.querySelector('span.lvl2')).toBeInTheDocument();
    });

    it('should have children', () => {
      const { container } = render(<CheckBox><div className="lvl1"><span className="lvl2">TEXT</span></div></CheckBox>);

      expect(container.children).toBeInstanceOf(HTMLCollection);

      expect(container.querySelector('div.lvl1 span.lvl2')?.textContent).toEqual('TEXT');
    });

    it('children should be passed to .checkbox-label', () => {
      const { container } = render(<CheckBox><div className="lvl1"><span className="lvl2">TEXT</span></div></CheckBox>);

      expect(container.querySelector('.checkbox-label div.lvl1 span.lvl2')).toBeInTheDocument();
    });
  });
});
