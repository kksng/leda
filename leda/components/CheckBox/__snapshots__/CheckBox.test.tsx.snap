// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox SNAPSHOTS should render 1`] = `
<div>
  <span>
    <input
      class="checkbox-input"
      id="test"
      type="checkbox"
    />
    <label
      class="checkbox-label"
      for="test"
    />
  </span>
</div>
`;

exports[`CheckBox SNAPSHOTS should render controllable mode 1`] = `
<div>
  <span>
    <input
      checked=""
      class="checkbox-input"
      id="test"
      name="checker"
      type="checkbox"
    />
    <label
      class="checkbox-label"
      for="test"
    />
  </span>
</div>
`;

exports[`CheckBox SNAPSHOTS should render controllable mode 2`] = `
<div>
  <span>
    <input
      checked=""
      class="checkbox-input"
      id="test"
      name="checker"
      type="checkbox"
    />
    <label
      class="checkbox-label"
      for="test"
    />
  </span>
</div>
`;
